src/lib/rule2E.ts

```diff
@@ -0,0 +1,85 @@
+import {Context, getDefaultContext} from './context';
+import {computeTextAlternative} from './compute_text_alternative';
+
+// Input types that imply role 'textbox'
+const TEXTBOX_INPUT_TYPES = ['email', 'tel', 'text', 'url'];
+
+/**
+ * Checks if input elem has role 'textbox' or has a list attribute
+ * @param elem - input element whose role is being calculated
+ * @return - true if elem has role 'textbox' or list attribute,
+ * false otherwise.
+ */
+function isTextboxOrListInput(elem: HTMLInputElement): boolean {
+  const inputType = elem.getAttribute('type')?.toLowerCase() ?? '';
+  return TEXTBOX_INPUT_TYPES.includes(inputType) || elem.hasAttribute('list');
+}
+
+// Input types that imply role 'range'
+const RANGE_INPUT_TYPES = ['number', 'range'];
+
+/**
+ * Checks if input elem has role 'range'
+ * @param elem - input element whose role is being calculated
+ * @return - true if elem has role 'range', false otherwise
+ */
+function isRangeInput(elem: HTMLInputElement): boolean {
+  const inputType = elem.getAttribute('type')?.toLowerCase() ?? '';
+  return RANGE_INPUT_TYPES.includes(inputType);
+}
+
+/**
+ * Implementation for rule 2E.
+ * @param node - node whose text alternative is being calculated
+ * @param context - additional information relevant to the computation of a text
+ * alternative for node.
+ * @return text alternative for 'node' if rule 2E accepts 'node', null otherwise.
+ */
+export function rule2E(
```

Comment:
> If the embedded control has role menu button, return the text alternative of the button.

Please add a comment explaining why this is missing (or implment it 🙂 )

Reply:
I think 2F handles this case, but I'll definitely add a comment explaining that!